### PR TITLE
Fixes to wazuh-db sync-agent-groups-get endpoint

### PR DESCRIPTION
**Pull request by sdvendramini**
|Related issues|Manual testing|Integration tests|
|---|---|---|
|https://github.com/wazuh/wazuh/issues/16188|https://github.com/wazuh/wazuh-qa/issues/3951|https://github.com/wazuh/wazuh-qa/pull/3959|
|https://github.com/wazuh/wazuh/issues/16190|-|-|
|https://github.com/wazuh/wazuh/issues/16196|-|-|

## Description

These changes have been made to fix the above issues. They are related to ```sync-agent-groups-get``` wazuh-db endpoint and consist in:
- New function created to add global group hash to ```sync-agent-groups-get``` response. 
- The ```condition``` parameter of ```sync-agent-groups-get``` is optional now. 
- Fix to not set ```group_sync_status``` to ```synced``` status when the response is full.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)

- [x] Working on cluster environments
- [x] Added unit tests (for new features)

### CODE REVIEW

**Comment by sdvendramini** [comment]
- https://github.com/wazuh/wazuh/issues/16188

This fix was tested on master quering the first chunck of data from the ```sync-agent-groups-get```:
```
python3 wdb-query.py 'global sync-agent-groups-get {"condition": "sync_status", "set_synced": true, "get_global_hash": true}'
due [{"data":[{"id":1,"groups":["g0","g1","g2"]},...,{"id":1739,"groups":["g0","g1","g2"]}]}]
```

Then we check that the last agent of the chuck is marked as ```synced``` and the next one still is marked as ```synreq```:
```
python3 wdb-query.py 'global sql select group_sync_status from agent where id=1739'
[
    {
        "group_sync_status": "synced"
    }
]

python3 wdb-query.py 'global sql select group_sync_status from agent where id=1740'
[
    {
        "group_sync_status": "syncreq"
    }
]
```
Finally we query the second chunck of agents and groups and check again that agents returned are correct.
```
root@jellyfish:~/wazuh-tools/utils# python3 wdb-query.py 'global sync-agent-groups-get {"condition": "sync_status", "set_synced": true, "get_global_hash": true}'
due [{"data":[{"id":1740,"groups":["g0","g1","g2"]},...,{"id":3449,"groups":["g0","g1","g2"]}]}]
```
We can observer once again that the last agent of the query and the following are marked correctly:
```
python3 wdb-query.py 'global sql select group_sync_status from agent where id=3449'
[
    {
        "group_sync_status": "synced"
    }
]

python3 wdb-query.py 'global sql select group_sync_status from agent where id=3450'
[
    {
        "group_sync_status": "syncreq"
    }
]
```

**Comment by sdvendramini** [comment]
- https://github.com/wazuh/wazuh/issues/16190

This fix was tested on worker. We can observe the following agents with theris groups:
```
python3 wdb-query.py 'global sql select id, "GROUP", group_hash from agent LIMIT 3'
[
    {
        "id": 0
    },
    {
        "id": 1,
        "group": "g0,g1,g2",
        "group_hash": "18832fd6"
    },
    {
        "id": 2,
        "group": "g0,g1,g2",
        "group_hash": "18832fd6"
    }
]
```

Then one of the groups is deleted:
```
root@focal:~# rm -r /var/ossec/etc/shared/g2

python3 wdb-query.py 'global sql select id, "GROUP", group_hash from agent LIMIT 3'
[
    {
        "id": 0
    },
    {
        "id": 1,
        "group": "g0,g1",
        "group_hash": "f37fe7e3"
    },
    {
        "id": 2,
        "group": "g0,g1",
        "group_hash": "f37fe7e3"
    }
]
```
After that we check if any agent is with ```group_sync_status="syncreq"```
```
python3 wdb-query.py 'global sql select count(id) from agent where id > 0 AND group_sync_status = "synreq"'
[
    {
        "count(id)": 0
    }
]

python3 wdb-query.py 'global sql select count(id) from agent where id > 0 AND group_sync_status = "synced"'
[
    {
        "count(id)": 25000
    }
]

```

Finally the master sync the agent group to the worker correctly:
```
python3 wdb-query.py 'global sql select id, "GROUP", group_hash from agent LIMIT 3'
[
    {
        "id": 0
    },
    {
        "id": 1,
        "group": "g0,g1,g2",
        "group_hash": "18832fd6"
    },
    {
        "id": 2,
        "group": "g0,g1,g2",
        "group_hash": "18832fd6"
    }
]

```

**Comment by sdvendramini** [comment]
- https://github.com/wazuh/wazuh/issues/16196

This change was tested quering ```sync-agent-groups-get``` with differents paremeters:
```
python3 wdb-query.py 'global sync-agent-groups-get {"get_global_hash": true}'
[
    {
        "data": [],
        "hash": "1945e0d9718edae6897a8c8777ad805a4fe6c9cd"
    }
]
```

```
python3 wdb-query.py 'global sync-agent-groups-get {}'
[
    {
        "data": []
    }
]
```

```
python3 wdb-query.py 'global sql update agent set group_sync_status = "syncreq" where id = 3000'
[]

python3 wdb-query.py 'global sync-agent-groups-get {"condition": "sync_status", "get_global_hash": true}'
[
    {
        "data": [
            {
                "id": 3000,
                "groups": [
                    "g0",
                    "g1",
                    "g2"
                ]
            }
        ],
        "hash": "1945e0d9718edae6897a8c8777ad805a4fe6c9cd"
    }
]
```

```
python3 wdb-query.py 'global sql update agent set group_sync_status = "syncreq" where id = 3000'
[]

python3 wdb-query.py 'global sync-agent-groups-get {"get_global_hash": true}'
[
    {
        "data": [],
        "hash": "1945e0d9718edae6897a8c8777ad805a4fe6c9cd"
    }
]
```

File: src/wazuh_db/wdb_parser.c [hunk 1]
```diff
@@ -5678,30 +5678,26 @@ int wdb_parse_global_sync_agent_groups_get(wdb_t* wdb, char* input, char* output
...
         } else {
-            wdb_groups_sync_condition_t condition = WDB_GROUP_INVALID_CONDITION;
+            wdb_groups_sync_condition_t condition = WDB_GROUP_NO_CONDITION;
             int last_id = 0;
             bool set_synced = false;
             bool get_hash = false;
             int agent_registration_delta = 0;
 
-            if (0 == strcmp(j_sync_condition->valuestring, "sync_status")) {
+            if (cJSON_IsString(j_sync_condition) && 0 == strcmp(j_sync_condition->valuestring, "sync_status")) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_parser.c:5698]:
```suggestion
            if (j_sync_condition && 0 == strcmp(j_sync_condition->valuestring, "sync_status")) {
```

File: src/wazuh_db/wdb_parser.c [hunk 2]
```diff
@@ -5678,30 +5678,26 @@ int wdb_parse_global_sync_agent_groups_get(wdb_t* wdb, char* input, char* output
...
                 condition = WDB_GROUP_SYNC_STATUS;
-            } else if (0 == strcmp(j_sync_condition->valuestring, "all")) {
+            } else if (cJSON_IsString(j_sync_condition) && 0 == strcmp(j_sync_condition->valuestring, "all")) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_parser.c:5700]:
```suggestion
            } else if (j_sync_condition && 0 == strcmp(j_sync_condition->valuestring, "all")) {
```

File: src/wazuh_db/wdb.h [hunk 3]
```diff
@@ -2020,6 +2021,16 @@ wdbc_result wdb_global_sync_agent_groups_get(wdb_t* wdb,
...
 
+/**
+ * @brief Add global group hash to JSON response.
+ *
+ * @param wdb The Global struct database.
+ * @param response JSON response to fill with global group hash.
+ * @param response_size Current size of JSON response.
+ * @return int result to represent if global hash has being added to JSON response.
+ */
+int wdb_global_add_global_group_hash_to_resposne(wdb_t *wdb, cJSON** response, size_t response_size);
```
Comment (TomasTurina) [inline src/wazuh_db/wdb.h:2032]:
```suggestion
int wdb_global_add_global_group_hash_to_response(wdb_t *wdb, cJSON** response, size_t response_size);
```

File: src/wazuh_db/wdb_global.c [hunk 4]
```diff
@@ -1473,8 +1473,7 @@ wdbc_result wdb_global_sync_agent_groups_get(wdb_t *wdb, wdb_groups_sync_conditi
             sync_statement_index = WDB_STMT_GLOBAL_GROUP_SYNC_ALL_GET;
             break;
         default:
-            mdebug1("Invalid groups sync condition");
-            return WDBC_ERROR;
+            break;
     }
 
     if (!wdb->transaction && wdb_begin2(wdb) < 0) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_global.c:1479]: I think this if can be inside the if below:
```
if (condition != WDB_GROUP_NO_CONDITION) {
    if (!wdb->transaction && wdb_begin2(wdb) < 0) {
        mdebug1("Cannot begin transaction");
        return WDBC_ERROR;
    }
...
```

File: src/wazuh_db/wdb_global.c [hunk 5]
```diff
@@ -1493,100 +1492,116 @@ wdbc_result wdb_global_sync_agent_groups_get(wdb_t *wdb, wdb_groups_sync_conditi
...
+            status = WDBC_OK;
         }
-        cJSON_Delete(j_agent_stmt);
     }
 
     return status;
 }
 
+int wdb_global_add_global_group_hash_to_resposne(wdb_t *wdb, cJSON** response, size_t response_size) {
+    if (response == NULL || *response == NULL) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_global.c:1582]:
```suggestion
    if (response == NULL || !cJSON_IsObject(*response)) {
```

**Review by TomasTurina — CHANGES_REQUESTED** [review]

**Review by TomasTurina — APPROVED** [review]

File: src/wazuh_db/wdb_global.c [hunk 6]
```diff
@@ -868,7 +868,7 @@ int wdb_global_delete_group(wdb_t *wdb, char* group_name) {
             cJSON* agent_id = cJSON_GetObjectItem(agent_id_item, "id_agent");
             if (cJSON_IsNumber(agent_id)) {
                 if (WDBC_ERROR == wdb_global_if_empty_set_default_agent_group(wdb, agent_id->valueint) ||
-                    WDBC_ERROR == wdb_global_recalculate_agent_groups_hash(wdb, agent_id->valueint, "syncreq")) {
+                    WDBC_ERROR == wdb_global_recalculate_agent_groups_hash(wdb, agent_id->valueint, wconfig.is_worker?"synced":"syncreq")) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_global.c:871]: Check also for standalone.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

**Review by TomasTurina — DISMISSED** [review]

**Comment by sdvendramini** [comment]
## Valgrind
For this test, two manager (1 master and 1 worker) were configured, 25.000 agents were enrolled and finally 4 groups were created and assigned to all the agents. Then some groups folder were deleted to force the synchronization process.

### Valgrind Reports 

- Master node
```
==69078== 
==69078== FILE DESCRIPTORS: 4 open (3 std) at exit.
==69078== Open file descriptor 3: /dev/urandom
==69078==    at 0x52A16EB: open (open64.c:41)
==69078==    by 0x23F401: randombytes (randombytes.c:62)
==69078==    by 0x23F536: os_random (randombytes.c:88)
==69078==    by 0x2354A1: OSHash_Create (hash_op.c:54)
==69078==    by 0x11D7CD: main (main.c:128)
==69078== 
==69078== 
==69078== HEAP SUMMARY:
==69078==     in use at exit: 0 bytes in 0 blocks
==69078==   total heap usage: 6,234,762 allocs, 6,234,762 frees, 496,521,914 bytes allocated
==69078== 
==69078== All heap blocks were freed -- no leaks are possible
==69078== 
==69078== For lists of detected and suppressed errors, rerun with: -s
==69078== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==69078== could not unlink /tmp/vgdb-pipe-from-vgdb-to-69078-by-root-on-???
==69078== could not unlink /tmp/vgdb-pipe-to-vgdb-from-69078-by-root-on-???
==69078== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-69078-by-root-on-???
```
- Worker node
```
==151736== 
==151736== FILE DESCRIPTORS: 5 open at exit.
==151736== Open AF_UNIX socket 5: queue/db/wdb
==151736==    at 0x51964FF: accept (accept.c:26)
==151736==    by 0x11E239: run_dealer (main.c:310)
==151736==    by 0x518B608: start_thread (pthread_create.c:477)
==151736==    by 0x52C5132: clone (clone.S:95)
==151736== 
==151736== Open file descriptor 3: /dev/urandom
==151736==    at 0x5196ABB: open (open64.c:48)
==151736==    by 0x23D36E: randombytes (randombytes.c:62)
==151736==    by 0x23D497: os_random (randombytes.c:88)
==151736==    by 0x2341AF: OSHash_Create (hash_op.c:54)
==151736==    by 0x11D76A: main (main.c:128)
==151736== 
==151736== Open file descriptor 2: /dev/pts/1
==151736==    <inherited from parent>
==151736== 
==151736== Open file descriptor 1: /dev/pts/1
==151736==    <inherited from parent>
==151736== 
==151736== Open file descriptor 0: /dev/pts/1
==151736==    <inherited from parent>
==151736== 
==151736== 
==151736== HEAP SUMMARY:
==151736==     in use at exit: 0 bytes in 0 blocks
==151736==   total heap usage: 22,884,884 allocs, 22,884,884 frees, 16,484,860,200 bytes allocated
==151736== 
==151736== All heap blocks were freed -- no leaks are possible
==151736== 
==151736== For lists of detected and suppressed errors, rerun with: -s
==151736== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==151736== could not unlink /tmp/vgdb-pipe-from-vgdb-to-151736-by-root-on-???
==151736== could not unlink /tmp/vgdb-pipe-to-vgdb-from-151736-by-root-on-???
==151736== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-151736-by-root-on-???
```

**Review by TomasTurina — APPROVED** [review]

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**